gooddata-sdk/gooddata_sdk/catalog/data_source/service.py

```diff
@@ -80,3 +86,59 @@ def register_upload_notification(self, data_source_id: str) -> None:
 
     def get_declarative_data_sources(self) -> CatalogDeclarativeDataSources:
         return CatalogDeclarativeDataSources.from_api(self._layout_api.get_data_sources_layout())
+
+    def put_declarative_data_sources(
+        self,
+        declarative_data_sources: CatalogDeclarativeDataSources,
+        credentials_path: Optional[Path] = None,
+        test_flag: bool = False,
+    ) -> None:
+        credentials: Optional[dict[str, Any]] = None
+        if credentials_path is not None:
+            with open(credentials_path, "r", encoding="utf-8") as f:
+                credentials = yaml.safe_load(f)["data_sources"]
+        if test_flag:
+            self._test_data_sources_connection(declarative_data_sources, credentials)
+        self._layout_api.put_data_sources_layout(declarative_data_sources.to_api(credentials))
+
+    def store_declarative_data_sources(self, path: Path) -> None:
+        data_sources = self.get_declarative_data_sources()
+        for data_source in data_sources.data_sources:
+            file_path = path / f"{data_source.id}.yaml"
+            with open(file_path, "w+", encoding="utf-8") as f:
+                yaml.safe_dump(data_source.to_api().to_dict(), f, indent=2, sort_keys=True)
+
+    def _test_data_sources_connection(
+        self, declarative_data_sources: CatalogDeclarativeDataSources, credentials: dict[str, Any] = None
+    ) -> None:
+        credentials = credentials if credentials is not None else dict()
+        for declarative_data_source in declarative_data_sources.data_sources:
+            if credentials.get(declarative_data_source.id) is not None:
+                if declarative_data_source.type == BIGQUERY_TYPE:
+                    token = TokenCredentialsFromFile.token_from_file(credentials[declarative_data_source.id])
+                    response = self._scan_api.test_data_source_definition(
+                        declarative_data_source.to_test_request(token=token)
+                    )
+                else:
+                    response = self._scan_api.test_data_source_definition(
+                        declarative_data_source.to_test_request(password=credentials[declarative_data_source.id])
+                    )
+            else:
+                response = self._scan_api.test_data_source_definition(declarative_data_source.to_test_request())
+            if not response.successful:
+                raise ValueError(f"Test connection to {declarative_data_source.id} was not successful.")
```

Comment:
I would expose also the reason for the error. 
response.error contains it.
The question is how to format "ugly errors". Try it and suggest the best solution.